### ISSUE TITLE
pytorch模型导出为Onnx模型后结果前后不一致

### ISSUE THREAD

**Issue by Autonomf** [wenet-e2e/wenet#1651]
**描述**
使用comformer模型训练了一个粤语语音识别模型，后面使用export_onnx_gpu.py将pytorch模型导出为Onnx模型，两个模型识别结果相差很大，Onnx模型结果明显要差，识别结果如下，只举了一例：

**Pytorch模型识别结果**
![image](https://user-images.githubusercontent.com/77525800/211283419-9bf4639d-275c-4723-8f62-226d3a1402ec.png)

**Onnx模型识别结果**
![image](https://user-images.githubusercontent.com/77525800/211283887-348c2bdd-710d-45e3-95ea-4b4cf19e7f3b.png)

其中Pytorch模型使用的平台的wenet/bin/recognize.py进行推理，而Onnx模型使用的是wenet/bin/recognize_onnx_gpu.py进行推理

**环境**
python==3.8.13
onnx==1.12.0
onnxruntime-gpu==1.13.1
pytorch==1.10.0+py3.8_cuda111
torchaudio==0.10.0

困恼许久的问题，不知道如何解决，期待回复！

**Comment by robin1001** [comment]
对齐一下 pytorch 和 onnx 导出时的参数是否一致。核心的参数有 chunk_size, num_left_chunks。

**Comment by Autonomf** [comment]
> 对齐一下 pytorch 和 onnx 导出时的参数是否一致。核心的参数有 chunk_size, num_left_chunks。

感谢回复！我现在是是离线导出encoder和decoder，导出的onnx模型和训练时候的pytorch模型的参数是一致的，并没有修改参数。另外除了尝试导出粤语识别模型，还以相同的方式导出过相同的训练过程的中文识别模型，而中文模型导出前后相差不大。

**Comment by xingchensong** [comment]
您好，模型参数一致，解码参数是否一致？比如上面提到的chunk_size, num_left_chunks

**Comment by LRY1994** [comment]
最后怎么解决呢？ chunk_size, num_left_chunks 现在叫做decoding_chunk_size, decoding_num_left_chunks

**Comment by Slyne** [comment]
> 最后怎么解决呢？ chunk_size, num_left_chunks 现在叫做decoding_chunk_size, decoding_num_left_chunks

贴一下你的命令行？以及现象。

**Comment by LRY1994** [comment]
> > 最后怎么解决呢？ chunk_size, num_left_chunks 现在叫做decoding_chunk_size, decoding_num_left_chunks
> 
> 贴一下你的命令行？以及现象。

pt运行脚本
```
decode_checkpoint=$dir/avg_30.pt
decoding_chunk_size=
  ctc_weight=0.3
  reverse_weight=0.5

  python wenet/bin/recognize.py --gpu 0 \
    --modes $decode_modes \
    --config $dir/train.yaml \
    --data_type $data_type \
    --test_data data/test/data.list \
    --checkpoint $decode_checkpoint \
    --beam_size 10 \
    --batch_size 16 \
    --ctc_weight $ctc_weight \
    --reverse_weight $reverse_weight \
    --result_dir $result_dir \
    ${decoding_chunk_size:+--decoding_chunk_size $decoding_chunk_size}
```

onnx导出脚本：

```
decoding_chunk_size=-1
num_decoding_left_chunks=-1
ctc_weight=0.3
reverse_weight=0.5

python3 wenet/bin/export_onnx_gpu.py \
    --config=$model_dir/train.yaml \
    --checkpoint=$model_dir/avg_30.pt \
    --cmvn_file=$model_dir/global_cmvn  \
    --beam_size 10 \
    --ctc_weight $ctc_weight \
    --reverse_weight $reverse_weight \
    --blank_penalty 0 \
    --decoding_chunk_size $decoding_chunk_size \
    --num_decoding_left_chunks $num_decoding_left_chunks \
    --output_onnx_dir=$export_dir 
```


导出过程有一些输出，不知是否有影响
`/data/linry/wenet/wenet/utils/mask.py:218: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  max_len = max_len if max_len > 0 else lengths.max().item()
/data/linry/wenet/wenet/transformer/embedding.py:100: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert offset + size <= self.max_len
/data/linry/wenet/wenet/transformer/attention.py:373: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if cache.size(0) > 0:
/data/linry/wenet/wenet/transformer/attention.py:151: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if mask.size(-1) > 0:  # time2 > 0
/data/linry/wenet/wenet/transformer/convolution.py:116: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if mask_pad.size(2) > 0:  # time > 0
/data/linry/wenet/wenet/transformer/convolution.py:120: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if cache.size(2) == 0:  # cache_t == 0
/data/linry/wenet/wenet/transformer/convolution.py:126: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert (x.size(2) > self.lorder)
/data/linry/wenet/wenet/transformer/convolution.py:147: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if mask_pad.size(2) > 0:  # time > 0
/data/linry/anaconda3/envs/wenet/lib/python3.8/site-packages/torch/onnx/_internal/jit_utils.py:307: UserWarning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied. (Triggered internally at ../torch/csrc/jit/passes/onnx/constant_fold.cpp:179.)
  _C._jit_pass_onnx_node_shape_type_inference(node, params_dict, opset_version)
/data/linry/anaconda3/envs/wenet/lib/python3.8/site-packages/torch/onnx/utils.py:702: UserWarning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied. (Triggered internally at ../torch/csrc/jit/passes/onnx/constant_fold.cpp:179.)
  _C._jit_pass_onnx_graph_shape_type_inference(
/data/linry/anaconda3/envs/wenet/lib/python3.8/site-packages/torch/onnx/utils.py:1209: UserWarning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied. (Triggered internally at ../torch/csrc/jit/passes/onnx/constant_fold.cpp:179.)
  _C._jit_pass_onnx_graph_shape_type_inference(
2024-06-24 21:34:35.951748111 [W:onnxruntime:, session_state.cc:1162 VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.
2024-06-24 21:34:35.951795112 [W:onnxruntime:, session_state.cc:1164 VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
/data3/linry/wenet/wenet/bin/export_onnx_gpu.py:739: FutureWarning: `torch.testing.assert_allclose()` is deprecated since 1.12 and will be removed in a future release. Please use `torch.testing.assert_close()` instead. You can find detailed upgrade instructions in https://github.com/pytorch/pytorch/issues/61844.
  torch.testing.assert_allclose(a, b, rtol=rtol, atol=atol)
Tensor-likes are not close!

Mismatched elements: 91507 / 393216 (23.3%)
Greatest absolute difference: 0.004305481910705566 at index (11, 12, 341) (up to 1e-05 allowed)
Greatest relative difference: 65.1364974975586 at index (24, 23, 251) (up to 0.001 allowed)
Tensor-likes are not close!

Mismatched elements: 9257 / 4573440 (0.2%)
Greatest absolute difference: 0.03157997131347656 at index (11, 12, 4639) (up to 1e-05 allowed)
Greatest relative difference: 0.025632504373788834 at index (11, 12, 0) (up to 0.001 allowed)
Tensor-likes are not close!

Mismatched elements: 229 / 7680 (3.0%)
Greatest absolute difference: 0.026508092880249023 at index (11, 12, 1) (up to 1e-05 allowed)
Greatest relative difference: 0.025632504373788834 at index (11, 12, 0) (up to 0.001 allowed)
Tensor-likes are not close!

Mismatched elements: 16 / 7680 (0.2%)
Greatest absolute difference: 4965 at index (14, 18, 6) (up to 1e-05 allowed)
Greatest relative difference: 7.280058860778809 at index (14, 18, 6) (up to 0.001 allowed)
INFO:/data3/linry/wenet/wenet/bin/export_onnx_gpu.py:export offline onnx encoder succeed!
/data/linry/wenet/wenet/transformer/attention.py:229: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if cache.size(0) > 0:
/data/linry/wenet/wenet/transformer/attention.py:469: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if cache.size(0) > 0:
/data/linry/wenet/wenet/transformer/attention.py:493: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if B != k.size(0):
/data/linry/wenet/wenet/transformer/attention.py:519: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!a
  if query.size(0) != B:
/data/linry/wenet/wenet/transformer/decoder.py:192: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  return x, torch.tensor(0.0), olens
/data/linry/wenet/wenet/transformer/decoder.py:426: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  r_x = torch.tensor(0.0)
2024-06-24 21:35:03.434509494 [W:onnxruntime:, session_state.cc:1162 VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.
2024-06-24 21:35:03.434956959 [W:onnxruntime:, session_state.cc:1164 VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
INFO:/data3/linry/wenet/wenet/bin/export_onnx_gpu.py:export to onnx decoder succeed!
`

**Comment by LRY1994** [comment]
我找到bug了！
wenet/runtime/gpu/model_repo/scoring/1/model.py
sos = eos = len(vocab) - 1  # 这里应该改为sos = eos =2

/wenet/bin/recognize_onnx_gpu.py
eos = sos =  len(char_dict) - 1 # 这里应该改为sos = eos =2

和这个  echo "<sos/eos> 2" >> $dict 矛盾
if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then
  echo "Make a dictionary"
  mkdir -p $(dirname $dict)
  echo "<blank> 0" > ${dict}  # 0 is for "blank" in CTC
  echo "<unk> 1"  >> ${dict}  # <unk> must be 1
  echo "<sos/eos> 2" >> $dict
  tools/text2token.py -s 1 -n 1 data/train/text | cut -f 2- -d" " \
    | tr " " "\n" | sort | uniq | grep -a -v -e '^\s*$' | \
    awk '{print $0 " " NR+2}' >> ${dict}
fi

**Comment by xingchensong** [comment]
> 我找到bug了！ wenet/runtime/gpu/model_repo/scoring/1/model.py sos = eos = len(vocab) - 1 # 这里应该改为sos = eos =2
> 
> /wenet/bin/recognize_onnx_gpu.py eos = sos = len(char_dict) - 1 # 这里应该改为sos = eos =2
> 
> 和这个 echo "<sos/eos> 2" >> $dict 矛盾 if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then echo "Make a dictionary" mkdir -p $(dirname $dict) echo " 0" > ${dict} # 0 is for "blank" in CTC echo " 1" >> ${dict} # must be 1 echo "<sos/eos> 2" >> $dict tools/text2token.py -s 1 -n 1 data/train/text | cut -f 2- -d" " | tr " " "\n" | sort | uniq | grep -a -v -e '^\s*$' | awk '{print Extra close brace or missing open brace $0 " " NR+2}' &gt;&gt; ${dict} fi

可以提个pr哈

**Comment by xingchensong** [comment]
@LRY1994 嗨请问方便提个pr吗
